google-cloud-datastore/src/main/java/com/google/cloud/datastore/spi/v1/GrpcDatastoreRpc.java

```diff
@@ -0,0 +1,211 @@
+    return datastoreStub.commitCallable().call(request);
+  }
+
+  @Override
+  public LookupResponse lookup(LookupRequest request) {
+    return datastoreStub.lookupCallable().call(request);
+  }
+
+  @Override
+  public ReserveIdsResponse reserveIds(ReserveIdsRequest request) {
+    return datastoreStub.reserveIdsCallable().call(request);
+  }
+
+  @Override
+  public RollbackResponse rollback(RollbackRequest request) {
+    return datastoreStub.rollbackCallable().call(request);
+  }
+
+  @Override
+  public RunQueryResponse runQuery(RunQueryRequest request) {
+    return datastoreStub.runQueryCallable().call(request);
+  }
+
+  @Override
+  public RunAggregationQueryResponse runAggregationQuery(RunAggregationQueryRequest request) {
+    return datastoreStub.runAggregationQueryCallable().call(request);
+  }
+
+  private boolean isEmulator(DatastoreOptions datastoreOptions) {
+    return datastoreOptions.getHost().contains("localhost")
+        || NoCredentials.getInstance().equals(datastoreOptions.getCredentials());
+  }
+
+  private ClientContext getClientContextForEmulator(DatastoreOptions datastoreOptions)
+      throws IOException {
+    ManagedChannel managedChannel =
+        ManagedChannelBuilder.forTarget(datastoreOptions.getHost()).usePlaintext().build();
+    TransportChannel transportChannel = GrpcTransportChannel.create(managedChannel);
+    return ClientContext.newBuilder()
+        .setCredentials(null)
+        .setTransportChannel(transportChannel)
+        .setDefaultCallContext(GrpcCallContext.of(managedChannel, CallOptions.DEFAULT))
+        .setBackgroundResources(Collections.singletonList(transportChannel))
+        .build();
+  }
+
+  private ClientContext getClientContext(DatastoreOptions datastoreOptions) throws IOException {
+    HeaderProvider internalHeaderProvider =
+        DatastoreSettings.defaultApiClientHeaderProviderBuilder()
+            .setClientLibToken(
+                ServiceOptions.getGoogApiClientLibName(),
+                GaxProperties.getLibraryVersion(datastoreOptions.getClass()))
+            .setResourceToken(getResourceToken(datastoreOptions))
+            .build();
+
+    DatastoreSettingsBuilder settingsBuilder =
+        new DatastoreSettingsBuilder(DatastoreSettings.newBuilder().build());
+    settingsBuilder.setCredentialsProvider(
+        GrpcTransportOptions.setUpCredentialsProvider(datastoreOptions));
+    settingsBuilder.setTransportChannelProvider(
+        GrpcTransportOptions.setUpChannelProvider(
+            DatastoreSettings.defaultGrpcTransportProviderBuilder(), datastoreOptions));
+    settingsBuilder.setInternalHeaderProvider(internalHeaderProvider);
+    settingsBuilder.setHeaderProvider(
+        datastoreOptions.getMergedHeaderProvider(new NoHeaderProvider()));
+    ClientContext clientContext = ClientContext.create(settingsBuilder.build());
+    return clientContext;
+  }
+
+  private String getResourceToken(DatastoreOptions datastoreOptions) {
```

Comment:
IIUC this shouldn't be needed anymore - can you check if we remove this if this is still present (via the generated code?)

Reply:
these are the headers which are being forwarded, out of which `google-cloud-resource-prefix` is what we are configuring.
![header_key](https://github.com/googleapis/java-datastore/assets/5915092/e6a8e492-ae84-48a8-b885-08edbddedd3c)

### toString representation 
`Metadata(x-goog-api-client=gl-java/17.0.9__Debian gccl/ gapic/ gax/2.37.0 grpc/1.59.0,google-cloud-resource-prefix=project_id=prod-env&database_id=primary-replica,grpc-accept-encoding=gzip,user-agent=grpc-java-inprocess/1.59.0,grpc-timeout=49966459u)`

Reply:
this code has been removed in the latest snapshot.

Reply:
in a followup PR (feel free to make an issue for this to track), we should add some tests around this)